### PR TITLE
Fix loopback issue injecting istio proxy

### LINKED ISSUE #103: connection errors with loopback http2 upgrade
PR #100 fails in "a" -> "a" case:
```
kubectl exec -it a-3708111894-q8dx0 client http://a/a
-- error
kubectl exec -it a-3708111894-q8dx0 client http://b/a
-- works fine
```
Here the debug log from envoy:
```
[2017-02-08 18:06:58.109][17][info][main] [C16] new connection
[2017-02-08 18:06:58.110][17][info][client] [C17] connecting
[2017-02-08 18:06:58.111][17][info][client] [C17] protocol error: The user callback function failed
[2017-02-08T18:06:58.110Z] "GET /a HTTP/1.1" 503 UC 0 57 2 - "-" "Go-http-client/1.1" "06c8f613-0618-408f-9a62-230286b91d2c" "a" "tcp://10.0.0.114:8080"
```

### PR DESCRIPTION
**Pull request by lizan**
Previously for loopback traffic the iptables rule routed it as:
app -> proxy -> app

This will route it as:
app -> proxy (client-side) -> proxy (server-side) -> app

So proxy can upgrade the protocol between proxies (e.g. HTTP/2, TLS)

cc @myidpt @kyessenov 

Closes #103 .

### CODE REVIEW

File: docker/prepare_proxy.sh [hunk 1]
```diff
@@ -41,6 +41,12 @@ if [[ -z "${ISTIO_PROXY_PORT-}" ]] || [[ -z "${ISTIO_PROXY_UID-}" ]]; then
 fi
 
 iptables -t nat -A PREROUTING -p tcp -j REDIRECT --to-port ${ISTIO_PROXY_PORT}
+# To make sure when pod A wants to talk to service A, which is backed by pod A,
+# the traffic is going through proxy twice, client-side and server-side.
+# lo traffic doesn't go through PREROUTING, so needed to be processed in OUTPUT.
```
Comment (kyessenov) [inline docker/prepare_proxy.sh]: can you expand on the meaning on the rule? e.g. concretely in terms of TCP quadruples and POD_IPs

File: docker/prepare_proxy.sh [hunk 2]
```diff
@@ -41,6 +41,12 @@ if [[ -z "${ISTIO_PROXY_PORT-}" ]] || [[ -z "${ISTIO_PROXY_UID-}" ]]; then
...
+iptables -t nat -A OUTPUT -p tcp -j REDIRECT -o lo \
+  ! -d 127.0.0.1/32 --to-port ${ISTIO_PROXY_PORT}
```
Comment (kyessenov) [inline docker/prepare_proxy.sh]: how is this not capturing traffic from proxy?

Comment (lizan) [inline docker/prepare_proxy.sh]: this is intended to capture traffic from (client-side) proxy to (server-side) proxy, and exclude (server-side) proxy to app by excluding `-d 127.0.0.1/32`.

**Comment by lizan** [comment]
Thanks @ayj! added comments based on yours to the shell script for each iptables rule.

**Pull request merged**